### PR TITLE
Catch the case when the last argument passed to n4jsc.jar is not a valid list of files

### PR DESCRIPTION
**Pull request by qtran-n4**
Task: https://github.com/eclipse/n4js/issues/794

### CODE REVIEW

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java [hunk 1]
```diff
@@ -381,6 +381,15 @@ public SuccessExitStatus doMain(String... args) throws ExitCodeException {
 				return SuccessExitStatus.INSTANCE;
 			}
 
+			// Make sure the srcFiles are valid
+			for (File srcFile : srcFiles) {
+				if (!srcFile.isFile() && !srcFile.isDirectory()) {
+					System.out.println(srcFile.toString()
```
Comment (yoosiba) [inline tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java]: `System.err`

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/ErrorExitCode.java [hunk 2]
```diff
@@ -42,7 +42,10 @@
 	/** Exit with 10, if required tester could not be loaded. */
 	EXITCODE_TESTER_NOT_FOUND(10, "required tester could not be loaded"),
 	/** Exit with 10, if required tester could not be loaded. */
-	EXITCODE_DEPENDENCY_NOT_FOUND(10, "required dependency is missing");
+	EXITCODE_DEPENDENCY_NOT_FOUND(10, "required dependency is missing"),
+	/** Exit with 11, if srcFiles are invalid. */
+	EXITCODE_SRCFILES_INVALID(10,
```
Comment (yoosiba) [inline tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/ErrorExitCode.java]: duplicated code `10`

Comment (yoosiba) [reply tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/ErrorExitCode.java]: sorry, triplicated, three exit codes with code `10`, and javadoc for two of them is wrong

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java [hunk 3]
```diff
@@ -381,6 +381,15 @@ public SuccessExitStatus doMain(String... args) throws ExitCodeException {
 				return SuccessExitStatus.INSTANCE;
 			}
 
+			// Make sure the srcFiles are valid
+			for (File srcFile : srcFiles) {
+				if (!srcFile.isFile() && !srcFile.isDirectory()) {
```
Comment (yoosiba) [inline tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java]: reuse `org.eclipse.n4js.hlc.base.HlcFileUtils.isExistingReadibleFile(File)`

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java [hunk 4]
```diff
@@ -381,6 +381,15 @@ public SuccessExitStatus doMain(String... args) throws ExitCodeException {
...
+					System.out.println(srcFile.toString()
+							+ " of the list of source files or projects is neither valid file nor valid directory.");
```
Comment (mor-n4) [inline tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java]: better move `srcFile.toString()` to the end of the message, separated by a colon

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java [hunk 5]
```diff
@@ -381,6 +381,15 @@ public SuccessExitStatus doMain(String... args) throws ExitCodeException {
...
+					throw new ExitCodeException(ErrorExitCode.EXITCODE_SRCFILES_INVALID);
```
Comment (mor-n4) [inline tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java]: consider checking all files before throwing the exception and then including all invalid files in the message

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java (hunk 3 above)
Comment (qtran-n4) [reply tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java]: Unfortunately, the method throws `RuntimeException` which is a bit ugly to reuse in this case...

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/ErrorExitCode.java (hunk 2 above)
Comment (qtran-n4) [reply tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/ErrorExitCode.java]: done

File: tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java (hunk 5 above)
Comment (qtran-n4) [reply tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/N4jscBase.java]: Done.

**Comment by qtran-n4** [comment]
@yoosiba Could you please have a second look and approve the PR if you are satisfied? :D

File: tests/org.eclipse.n4js.hlc.tests/src/org/eclipse/n4js/hlc/tests/N4jscBasicTest.java [hunk 6]
```diff
@@ -204,4 +206,24 @@ public void testCleanAllProjects() throws ExitCodeException {
...
+	@Test
+	public void testInvalidSrcFiles() {
+		String proot = workspace.getAbsolutePath().toString();
+		// Project
+		String pathToBLAH = "/BLAH";
+		String[] args = { "-pl", proot,
+				"--buildType", "projects", pathToBLAH,
+				"--verbose"
+		};
+		try {
```
Comment (yoosiba) [inline tests/org.eclipse.n4js.hlc.tests/src/org/eclipse/n4js/hlc/tests/N4jscBasicTest.java]: use ` org.eclipse.n4js.hlc.tests.AbstractN4jscTest.expectCompilerException(String[] args, ErrorExitCode expectedExitCode)`

**Comment by yoosiba** [comment]
@qtran-n4 did this go through builds? can you merge it?

**Comment by qtran-n4** [comment]
Still waiting for the final builds (should be green since I only made minimal changes). Will merge when the builds are done.

**Comment by qtran-n4** [comment]
Thanks for the reviews! Builds are green. Merging now.

**Pull request merged**